src/amulware.Graphics/Core/shading/CachedVariableLocator.cs

```diff
@@ -0,0 +1,27 @@
+using System;
+using System.Collections.Generic;
+
+namespace amulware.Graphics
+{
+    internal class CachedVariableLocator
```

Comment:
Isn't this just a LazyDictionary? Or put even more simply, a cache keyed by string. It sounds as if you called it after what it's used for, rather than what it is.

I am not even sure why this warrants such a specific class. As far as I could tell this could be solved with an extension method on a dictionary. The only reason I could think of using a class is to hide the implementation details, though there's no interface and the `CachedVariableLocator` as name kind of gives it away anyway.

Reply:
You are partially correct, this could easily be refactored into a `FunctionCache` class or similar. I don't think you should look at it as a dictionary though, that's an implementation detail. It's a cache decorator around a function.

I thought about creating such a class and maybe even adding it to utilities, but for the time being, this was the least disruptive and most semantic implementation, so I chose to go with this.

I'd be happy to see a refactor of this in the future though.

Note: It would not work as an extension of dictionary unless you want to pass in the fetch function on every call, which I think is inelegant and not semantically correct (see above about this not being a dictionary).